### ISSUE TITLE
Wrong constant long number evaluation in compiled code

### ISSUE THREAD

**Issue by vegegoku** [gwtproject/gwt#9708]
**GWT version:2.9.0** 
**Browser (with version):Chrome Version 85.0.4183.83 (Official Build) (64-bit)** 
**Operating System: Ubuntu 20.04 LTS** 

---

##### Description
Defining a constant long number like the following will produce the correct result in development mode but the wrong result in production mode : 
```java

public class App implements EntryPoint {

	private final static long MONTH = 31 * 24 * 60 * 60 * 1000;
	public void onModuleLoad() {
		DomGlobal.console.info(""+MONTH);
	}
}
```

in development mode, the result of the constant will be `2678400000` but if we run a compiled app the result will be `-1616567296` which is wrong.

##### Steps to reproduce

the entry point from the code snippet above all what we need to reproduce the bug

- Run the application with SDM and see the result, should be correct.
- Compile the app and run compiled, the result is wrong.

##### Known workarounds
if we add an `L` to any of the numbers from the expression the compiled result will be correct.

e.g 
`private final static long MONTH = 31 * 24 * 60 * 60 * 1000L;`

**Comment by rluble** [comment]
> in development mode, the result of the constant will be `2678400000` but if we run a compiled app the result will be `-1616567296` which is wrong.

It is actually the other way around. The correct result is `-1616567296` since all the operands are integer and according to Java semantics that implies the multiplication should be done with integer precision, resulting in overflow. In optimized mode, the compiler evaluates statically the expression according to Java semantics. But in draft mode the expression is left for JavaScript to evaluate which gives the seemingly correct but semantically incorrect result.

Integer operations in GWT don't fully follow Java semantics w.r.t overflows.

**Comment by vegegoku** [comment]
@rluble Yes, I see now, that's true, but it was confusing because of the inconsistency between the SDM and production mode.
